torch/_guards.py

```diff
@@ -781,6 +782,15 @@ def compile_context(context: Optional[CompileContext]):
     try:
         yield context
     finally:
+        if context is not None:
+            if context.compile_id is not None:
+                set_context_frame(
```

Comment:
(btw this feature is awesome 😄 )

side car comment, maybe cc @anijain2305 - won't this function get called (and set the context) during compilation? It seems like we would want this state to be set to the "previously compiled frame" **every time** a given compiled object is invoked at runtime.

Reply:
hmm yeah, repro here:

```
@torch.compile(backend="aot_eager", dynamic=True)
def f(x):
    if x.shape[0] > 5:
        return x.sin()
    else:
        return x.cos()

x1 = torch.randn(4)
x2 = torch.randn(6)
x3 = torch.randn(3)

with torch.profiler.profile(record_shapes=True) as prof:
    out1 = f(x1)
    out2 = f(x2)
    out3 = f(x3)

print([e.kwinputs['context'] for e in prof.events() if 'Compiled' in e.name])
```

this should print:
```
['0/0', '0/1', '0/0']
```

since we are re-running the first compiled object on the third case. But instead, I get:
```
['0/0', '0/1', '0/1']
```

I can file an issue

Reply:
https://github.com/pytorch/pytorch/issues/136235